[Gloo] update gloo submodule for PyTorch

jiayisuse opened the pull request:
Summary: To include alltoall

Test Plan: CI

Differential Revision: D22544255

osalpekar commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -203,6 +205,16 @@ void setInput(O& opts, at::Tensor& tensor) {
 }
 
+template <typename T, typename O>
+void setInput(O& opts, at::Tensor& tensor, std::vector<size_t>& counts) {
```
Is the only difference between this and the `setInput` on Line 214 the data type in the `counts` vector? Do we need separate ones for `size_t` and `int64_t`?

jiayisuse replied:
Yes, alltoall interfaces use int64_t instead of size_t
https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroup.hpp#L171-L175

osalpekar commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2236,6 +2253,83 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::reduce_scatter(
+  std::vector<int64_t> inputCounts;
+  const uint32_t tag;
+
+  void run() override {
```
Seems like the coding style for the other `AsyncWork` classes is to implement all this logic in a helper function like `all_to_all(inputs, outputs)` and then simply call this helper from `run`. Should we use that same style here?

jiayisuse replied:
Yes, this AsyncAlltoallWork handles both `alltoall` and `alltoallv`. That's why we didn't create a separate function

osalpekar reviewed (approved):
Added a few comments, but stamping to unblock.

jiayisuse committed:
[Gloo] update gloo submodule for PyTorch (#41462)

Summary:
To include alltoall

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41462

Test Plan: CI

Reviewed By: osalpekar

Differential Revision: D22544255

Pulled By: jiayisuse

fbshipit-source-id: 249f9c4d2698c41c2b3afe0b6112ff7224968537
